### PR TITLE
Adding support for concurrency formula as an option in Optuna search

### PR DESCRIPTION
**Pull request by nv-braf**
Gives the user the option to use our concurrency formula rather than having Optuna search the full concurrency range.

By default we allow Optuna to search the concurrency space. This gives the user the option of using our concurrency formula (2 * instance_count * batch_size) instead, which will reduce the search space and should improve the search time.

### CODE REVIEW

File: tests/test_cli.py [hunk 1]
```diff
@@ -385,6 +386,7 @@ def _test_boolean_option(self, option_struct):
         cli = option_struct.cli_subcommand()
         _, config = cli.parse()
         option_value = config.get_config().get(option_with_underscores).value()
+        # Boolean values must always default to False
```
Comment (debermudez) [inline tests/test_cli.py:389]: why is this necessary?
The arg could use default=True

File: model_analyzer/config/generate/search_parameters.py [hunk 2]
```diff
@@ -90,23 +90,14 @@ def _number_of_configurations_for_parameter(
         return number_of_parameter_configs
 
     def _populate_search_parameters(self) -> None:
-        if self._parameters:
-            self._populate_parameters()
-        else:
-            self._populate_default_parameters()
```
Comment (debermudez) [inline model_analyzer/config/generate/search_parameters.py:96]: why did this method get removed?

File: tests/test_optuna_run_config_generator.py [hunk 3]
```diff
@@ -159,6 +159,43 @@ def test_create_objective_based_run_config(self):
...
+        trial = rcg._study.ask()
+        trial_objectives = rcg._create_trial_objectives(trial)
+        run_config = rcg._create_objective_based_run_config(trial_objectives)
+
+        model_config = run_config.model_run_configs()[0].model_config()
+        perf_config = run_config.model_run_configs()[0].perf_config()
+
+        self.assertEqual(model_config.to_dict()["name"], self._test_config_dict["name"])
+
+        # These values are the result of using a fixed seed of 100
```
Comment (debermudez) [inline tests/test_optuna_run_config_generator.py:189]: Nice note!

File: tests/test_search_parameters.py [hunk 4]
```diff
@@ -248,6 +248,38 @@ def test_search_parameter_creation_default(self):
...
+
+        yaml_content = """
+        profile_models: add_sub
+        """
+        config = TestConfig()._evaluate_config(args=args, yaml_content=yaml_content)
+
+        analyzer = Analyzer(config, MagicMock(), MagicMock(), MagicMock())
+        analyzer._populate_search_parameters()
+
+        concurrency = analyzer._search_parameters["add_sub"].get_parameter(
```
Comment (debermudez) [inline tests/test_search_parameters.py:277]: This wont error out?
get_parameter is designed to return None when a parameter isnt found?

File: tests/test_cli.py (hunk 1 above)
Comment (nv-braf) [reply tests/test_cli.py:389]: This is a byproduct of how the `ConfigPrimitive(bool)` works. The field must be False by default and then it is set to True by setting it on the CLI (or in the config). You cannot have a default of True (and set it to False via the CLI/config)

File: model_analyzer/config/generate/search_parameters.py (hunk 2 above)
Comment (nv-braf) [reply model_analyzer/config/generate/search_parameters.py:96]: This was just a placeholder until this code got added. The new code has the correct functionality.

File: tests/test_search_parameters.py (hunk 4 above)
Comment (nv-braf) [reply tests/test_search_parameters.py:277]: Yes, it will return None when the key is not found.

File: tests/test_cli.py (hunk 1 above)
Comment (debermudez) [reply tests/test_cli.py:389]: i see. okay

**Review by debermudez — APPROVED** [review]
